src/main/java/de/tum/in/www1/artemis/service/CourseService.java

```diff
@@ -333,8 +371,9 @@ public void cleanupCourse(Long courseId) {
         }
 
         // Clean up exams
-        var exams = courseRepository.findByIdWithLecturesAndExamsElseThrow(course.getId()).getExams();
-        var examExercises = exams.stream().map(exam -> examRepository.findAllExercisesByExamId(exam.getId())).flatMap(Collection::stream).collect(Collectors.toSet());
+        Set<Exercise> examExercises = examRepository.findByCourseIdWithExerciseGroupsAndExercises(courseId).stream().filter(Objects::nonNull).map(Exam::getExerciseGroups)
```

Comment:
Will the Database ever deliver a collection containing `null`?

Reply:
Removed some of them and added explaining comment in 1a865f7967521c67505bb647620907feb44d005a